File: backend/app/models/challenge.py
from sqlmodel import SQLModel, Field
from datetime import datetime
from app.schemas.challenge_schema import ChallengeType, ChallengeMode

class Challenge(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    owner: int | None = Field(default=None, foreign_key="user.id")
    name: str
    description: str
    xp_reward: int
    start_date: datetime
    end_date: datetime
    goal: int
    visibility: bool = False
    category: str
    type_challenge: ChallengeType
    mode_challenge: ChallengeMode
    max_participants: int 
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)

File: frontend/gym/src/schemas/challenge.ts
import z from "zod";

export const ChallengeSchema = z.object({
    id: z.number(),
    owner: z.number(),
    name: z.string(),
    description: z.string(),
    xp_reward: z.number(),
    start_date: z.coerce.date(),
    end_date: z.coerce.date(),
    goal: z.number(),
    visibility: z.boolean(),
    type_challenge: z.string(),
    category: z.string(),
});

export type Challenge = z.infer<typeof ChallengeSchema>;

export const CreateChallengeSchema = z.object({
    name: z.string(),
    description: z.string(),
    start_date: z.coerce.date(),
    end_date: z.coerce.date(),
    goal: z.number(),
    visibility: z.boolean(),
    type_challenge: z.string(),
    category: z.string(),
});

export type CreateChallenge = z.infer<typeof CreateChallengeSchema>;
